esrally/client.py

```diff
@@ -127,18 +128,39 @@ def _is_set(self, client_opts, k):
         except KeyError:
             return False
 
-    def create(self):
-        import elasticsearch
+    def _postprocess_client_options(self):
+        """
+        This method is used to do any non init processing of the client options, prior to creating a client
+        """
+        if self._is_set(self.client_options, "use_api_key"):
+            self._generate_api_key()
+
+    def _generate_api_key(self):
+        with self._create_sync_client() as instance:
+            # an api_key is generated per client, which happens after an initial handshake using an existing auth scheme
+            # once this key is generated, a new client is created using the api_key client option and http_auth is removed
+            # the reason the options were not removed in place was because of the logic to coalesce the key tuple into a
+            # header that the client could use.
+            api_key_response = instance.security.create_api_key({"name": "rally-api-key"})
+            self.client_options.pop("http_auth", None)
+            self.client_options["api_key"] = (api_key_response["id"], api_key_response["api_key"])
+
+    def _create_sync_client(self):
         return elasticsearch.Elasticsearch(hosts=self.hosts, ssl_context=self.ssl_context, **self.client_options)
 
+    def create(self):
+        self._postprocess_client_options()
```

Comment:
There are two issues with this that I missed when first eyeballing your draft PR:

1. This circumvents the logic to wait for the REST layer to be available. This will work if you spin up Elasticsearch yourself because by the time you start Rally, Elasticsearch is likely up, but it won't work when Rally spins up the cluster. The reason is that we first create the client, then use this client to wait for the REST API layer and only then continue with the benchmark (see `Driver#prepare_benchmark()` for details). Reproduction scenario: `esrally --distribution-version=7.9.3 --car=defaults,trial-license,x-pack-security --client-options="use_ssl:true,verify_certs:false,basic_auth_user:'rally',basic_auth_password:'rally-password',use_api_key:true"`
2. This will only create one API key per process but not per simulated client. Say, you start a bulk indexing task on an 8 core machine with 256 clients. Rally will start 8 load generator processes (because of the 8 cores) and each of them will retrieve an API key. Each of the load generator processes will simulate 256/8 = 32 clients and each of these clients will use the same API key. However, a user will likely expect that each client uses their own API key (i.e. retrieve 256 API keys).

Given these two issues I sense that we need a different approach. One option would be to insert a dedicated task at the beginning of a schedule but I sense that's the wrong approach either because the track should be agnostic to how clients authenticate. Maybe let's discuss this on a high-bandwidth medium.